### DISCUSSION TITLE
Update x-text when property of x-data object changes

### DISCUSSION THREAD

**Discussion by cybardev** [alpinejs/alpine#3608] [1. Help] [answered]
## Code

`CodePen`: [`cybarspace/LYgwjLr`](https://codepen.io/cybarspace/pen/LYgwjLr)

```html
<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="UTF-8" />
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <title>Counter</title>
        <link rel="stylesheet" href="https://unpkg.com/mvp.css@1.12/mvp.css" />
        <script
            defer
            src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js"
        ></script>
        <script>
            const data = {
                count: 0,
                counter() {
                    console.log(data.count++);
                    setTimeout(data.counter, 1000);
                },
            };
        </script>
    </head>
    <body>
        <section x-data="data">
            <aside x-init="counter">
                <p>Count: <span x-text="count"></span></p>
            </aside>
        </section>
    </body>
</html>
```

## Goal

- [x] `x-init` starts counter
- [x] `data.count` updates every second
- [x] webpage shows updated value

## Problem

Value of `data.count` changes (_verified by console log_) but is not reflected in the webpage.

## Reason

There's probably much better ways to implement an automatic incremental counter. This is a simplified model of updating `x-text` based on the results of an API call that modifies the properties of the `x-data` object.

## Request

Any help regarding "how to update `x-text` when object property changes" would be much appreciated.

**Comment by ekwoka** [comment]
X data accepts a function that returns an object.

**Reply by cybardev** [reply]
I've always used an object directly, but never had to deal with automatic updates until now (it was always handled by an event).

I just tried it in the CodePen and it seems to not work. I did this:

```js
const data = {
  count: 0,
  counter() {
    console.log(data.count++);
    setTimeout(data.counter, 1000);
  }
};

function getData() {
  return data;
}
```
It has the same behaviour - updates property but not DOM.

---

I also tried:
```js
function getData() {
  return {
    count: 0,
    counter() {
      console.log(this.count++);
      setTimeout(this.counter, 1000);
    }
  };
}
```
But that no longer updated the `count` property, even in console logs.

---

**PS**: In both cases I set `x-data` to `getData`.

**Reply by ekwoka** [reply]
The problem is you're modifying the underlying object and not the reactive context.

Here you could just make it be `this` instead of `data`.

But fundamentally, the issue is that you're modifying the wrong object.

That last one you show does work, you just need to write the correct properties.

**Reply by cybardev** [reply]
Ok, here's what I tried again on CodePen:

`HTML`:
```html
<section x-data="getData">
  <aside>
    <p>Count: <span x-text="count"></span></p>
  </aside>
</section>
<script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js"></script>
```
`JS`:
```js
function getData() {
  return {
    count: 0,
    counter() {
      console.log(this.count++);
      setTimeout(this.counter, 1000);
    },
    init() {
      this.counter();
    }
  };
}
```
`CSS`:
```css
@import url("https://unpkg.com/mvp.css@1.12/mvp.css");
```

Here's what I see:

<img width="1632" alt="image" src="https://github.com/alpinejs/alpine/assets/50134239/a9c3f649-36d3-4896-b5d3-3e658591dae4">

For some reason, the `count` property becomes `NaN` and then can no longer be incremented.

**Reply by cybardev** [reply]
Also, notice that in my separate comment below, I only had to use 1 line of raw JS to make it work. Just set the `id` of the `<span>`, and set `innerText` each iteration. However, that isn't an elegant solution.

---

> modifying the underlying object and not the reactive context

How to access the reactive context?

> need to write the correct properties

Which would that be?

---

I am just very confused... 😓

**Reply by ekwoka** [reply]
That is really strange.

What if you just do

`this.count += 1`
`this.count = this.count+1`

**Reply by ekwoka** [reply]
Oh

It's obviously 😂

You are breaking the this context on the set timeout.

You need to use an arrow function, or bind the this on the function.

Otherwise `this.count` becomes undefined.

Common mistake once people start using this. Even I didn't notice it for a bit despite definitely running into that and knowing about it.

**Reply by cybardev** [reply]
OH. Did not know that... Never used `setTimeout` before either. 😓 Now it all works regardless of how I initialize `x-data`, just by using a lambda in `setTimeout`.

Thanks a lot for your patient help. Much appreciated. 😊

**Comment by cybardev** [comment]
## Potentially Non-ideal Solution

This VanillaJS solution is not ideal, unless there is no way to do this in Alpine, in which case it's fine.

```html
<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="UTF-8" />
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <title>Counter</title>
        <link rel="stylesheet" href="https://unpkg.com/mvp.css@1.12/mvp.css" />
        <script
            defer
            src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js"
        ></script>
        <script>
            const data = {
                count: 0,
                counter() {
                    console.log(data.count++);
                    document.querySelector("#count").innerText = data.count;  // set innerText
                    setTimeout(data.counter, 1000);
                },
            };
        </script>
    </head>
    <body>
        <section x-data="data">
            <aside x-init="counter">
                <p>Count: <span id="count">0</span></p>  <!-- set id attribute -->
            </aside>
        </section>
    </body>
</html>
```

**Comment by cybardev** [comment, marked as answer]
## Ideal Solution

```html
<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="UTF-8" />
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <title>Counter</title>
        <link rel="stylesheet" href="https://unpkg.com/mvp.css@1.12/mvp.css" />
        <script
            defer
            src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js"
        ></script>
    </head>
    <body>
        <section>
            <aside x-data="data">
                <p>Count: <span x-text="count"></span></p>
            </aside>
        </section>
        <script>
            const data = {
                count: 0,
                counter() {
                    console.log(this.count++);
                    setTimeout(() => this.counter(), 1000);
                },
                init() {
                    this.counter();
                },
            };
        </script>
    </body>
</html>
```

### Credit

See @ekwoka 's [comment above](https://github.com/alpinejs/alpine/discussions/3608#discussioncomment-6074222).
